File: go/deploykit_test.go
package deploykit

import (
	"reflect"
	"strings"
	"testing"
)

func TestNewConfigDefaults(t *testing.T) {
	cfg := NewConfig()
	if cfg.AppName != "ledgerrag" {
		t.Errorf("AppName = %q, want ledgerrag", cfg.AppName)
	}
	if cfg.Image != "ghcr.io/example/ledgerrag:latest" {
		t.Errorf("Image = %q", cfg.Image)
	}
	if cfg.Replicas != 2 {
		t.Errorf("Replicas = %d, want 2", cfg.Replicas)
	}
	if !cfg.TLS {
		t.Error("TLS should default true")
	}
	if cfg.SecretBackend != "auto" {
		t.Errorf("SecretBackend = %q, want auto", cfg.SecretBackend)
	}
	if !cfg.DenyAllEgress {
		t.Error("DenyAllEgress should default true")
	}
}

func TestAllTargetsProducePlans(t *testing.T) {
	cfg := NewConfig()
	for _, name := range []string{"local", "azure", "aws"} {
		target, err := GetTarget(name)
		if err != nil {
			t.Fatalf("GetTarget(%q) error: %v", name, err)
		}
		plan := target.Plan(cfg)
		if len(plan.Steps) == 0 {
			t.Errorf("%s: expected non-empty steps", name)
		}
		if len(plan.Manifests) == 0 {
			t.Errorf("%s: expected manifests", name)
		}
		if !plan.IsSecure() {
			t.Errorf("%s: defaults should be secure", name)
		}
		if plan.Target != name {
			t.Errorf("%s: plan.Target = %q", name, plan.Target)
		}
	}
}

func TestSecureDefaultsPass(t *testing.T) {
	target, _ := GetTarget("azure")
	plan := target.Plan(NewConfig())
	if !plan.IsSecure() {
		t.Error("expected secure plan")
	}
	if len(plan.Warnings) != 0 {
		t.Errorf("expected no warnings, got %v", plan.Warnings)
	}
}

func TestInsecureConfigFlagged(t *testing.T) {
	cfg := NewConfig()
	cfg.TLS = false
	cfg.DenyAllEgress = false
	target, _ := GetTarget("aws")
	plan := target.Plan(cfg)
	if plan.IsSecure() {
		t.Error("expected insecure plan")
	}
	joined := strings.Join(plan.Warnings, "\n")
	if !strings.Contains(joined, "TLS") {
		t.Errorf("expected TLS warning, got %v", plan.Warnings)
	}
	if !strings.Contains(joined, "egress") {
		t.Errorf("expected egress warning, got %v", plan.Warnings)
	}
}

func TestPlaintextSecretBackendFlagged(t *testing.T) {
	cfg := NewConfig()
	cfg.SecretBackend = "plaintext"
	target, _ := GetTarget("local")
	plan := target.Plan(cfg)
	if plan.IsSecure() {
		t.Error("plaintext backend should be insecure")
	}
	if len(plan.Warnings) != 1 || !strings.Contains(plan.Warnings[0], "plaintext") {
		t.Errorf("expected plaintext warning, got %v", plan.Warnings)
	}
}

func TestWarningOrdering(t *testing.T) {
	cfg := NewConfig()
	cfg.TLS = false
	cfg.DenyAllEgress = false
	cfg.SecretBackend = "plaintext"
	target, _ := GetTarget("local")
	got := target.Plan(cfg).Warnings
	want := []string{
		"TLS disabled -- traffic would be unencrypted",
		"egress not restricted -- data exfil risk",
		"secrets stored in plaintext",
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("warnings = %v, want %v", got, want)
	}
}

func TestTargetSelectsRightSecretBackend(t *testing.T) {
	azure, _ := GetTarget("azure")
	if !strings.Contains(azure.Plan(NewConfig()).Manifests["main.bicep"], "keyvault") {
		t.Error("azure should use keyvault")
	}
	aws, _ := GetTarget("aws")
	if !strings.Contains(strings.Join(aws.Plan(NewConfig()).Steps, "\n"), "secrets-manager") {
		t.Error("aws should use secrets-manager")
	}
	local, _ := GetTarget("local")
	if !strings.Contains(local.Plan(NewConfig()).Manifests["values.yaml"], "k8s-secret") {
		t.Error("local should use k8s-secret")
	}
}

func TestNonAutoSecretBackendOverride(t *testing.T) {
	cfg := NewConfig()
	cfg.SecretBackend = "vault"
	local, _ := GetTarget("local")
	if !strings.Contains(local.Plan(cfg).Manifests["values.yaml"], "secretBackend: vault") {
		t.Error("override should propagate to manifest")
	}
}

func TestUnknownTargetError(t *testing.T) {
	_, err := GetTarget("gcp")
	if err == nil {
		t.Fatal("expected error for unknown target")
	}
	want := "unknown target 'gcp'. available: ['local', 'azure', 'aws']"
	if err.Error() != want {
		t.Errorf("error = %q, want %q", err.Error(), want)
	}
}

func TestHelmValuesExact(t *testing.T) {
	local, _ := GetTarget("local")
	got := local.Plan(NewConfig()).Manifests["values.yaml"]
	want := "image: ghcr.io/example/ledgerrag:latest\nreplicas: 2\ntls: true\nsecretBackend: k8s-secret\nnetworkPolicy:\n  denyAllEgress: true\n"
	if got != want {
		t.Errorf("values.yaml =\n%q\nwant\n%q", got, want)
	}
}

func TestBicepExact(t *testing.T) {
	azure, _ := GetTarget("azure")
	got := azure.Plan(NewConfig()).Manifests["main.bicep"]
	want := "// Bicep for ledgerrag\nparam image string = 'ghcr.io/example/ledgerrag:latest'\nparam secretBackend string = 'keyvault'\n"
	if got != want {
		t.Errorf("main.bicep =\n%q\nwant\n%q", got, want)
	}
}

func TestTerraformExact(t *testing.T) {
	aws, _ := GetTarget("aws")
	got := aws.Plan(NewConfig()).Manifests["main.tf"]
	want := "# Terraform for ledgerrag\nvariable \"image\" { default = \"ghcr.io/example/ledgerrag:latest\" }\n# secrets: secrets-manager\n"
	if got != want {
		t.Errorf("main.tf =\n%q\nwant\n%q", got, want)
	}
}

func TestTLSOffChangesIngressAndFrontSteps(t *testing.T) {
	cfg := NewConfig()
	cfg.TLS = false
	azure, _ := GetTarget("azure")
	azureSteps := strings.Join(azure.Plan(cfg).Steps, "\n")
	if strings.Contains(azureSteps, "with TLS") {
		t.Error("azure ingress step should drop TLS")
	}
	if !strings.Contains(azureSteps, "enable ingress") {
		t.Error("azure should still enable ingress")
	}
	aws, _ := GetTarget("aws")
	awsSteps := strings.Join(aws.Plan(cfg).Steps, "\n")
	if strings.Contains(awsSteps, "ACM TLS cert") {
		t.Error("aws should drop ACM TLS cert")
	}
	if !strings.Contains(awsSteps, "front with ALB") {
		t.Error("aws should still front with ALB")
	}
}

func TestDenyEgressOffRemovesStep(t *testing.T) {
	cfg := NewConfig()
	cfg.DenyAllEgress = false
	local, _ := GetTarget("local")
	steps := local.Plan(cfg).Steps
	if len(steps) != 4 {
		t.Errorf("expected 4 steps without egress lock, got %d", len(steps))
	}
	for _, s := range steps {
		if strings.Contains(s, "NetworkPolicy") {
			t.Error("egress-off plan should not add NetworkPolicy step")
		}
	}
}

func TestLocalStepsExactWithDefaults(t *testing.T) {
	local, _ := GetTarget("local")
	got := local.Plan(NewConfig()).Steps
	want := []string{
		"ensure kind/minikube cluster is running",
		"create namespace ledgerrag",
		"create k8s-secret for app credentials",
		"helm install ledgerrag ./charts/ledgerrag --set image=ghcr.io/example/ledgerrag:latest",
		"apply default-deny NetworkPolicy",
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("steps = %v, want %v", got, want)
	}
}

func TestRenderPlanSecureReturnsZero(t *testing.T) {
	text, code, err := RenderPlan("local", NewConfig())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if code != 0 {
		t.Errorf("code = %d, want 0", code)
	}
	if !strings.Contains(text, "secure-by-default checks: PASS") {
		t.Error("expected PASS line")
	}
	if !strings.HasPrefix(text, "== DeployKit plan: local / ledgerrag ==\n") {
		t.Errorf("unexpected header: %q", text)
	}
}

func TestRenderPlanInsecureReturnsOne(t *testing.T) {
	cfg := NewConfig()
	cfg.TLS = false
	cfg.DenyAllEgress = false
	text, code, err := RenderPlan("aws", cfg)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if code != 1 {
		t.Errorf("code = %d, want 1", code)
	}
	if !strings.Contains(text, "SECURITY WARNINGS:") {
		t.Error("expected warnings section")
	}
}

func TestRenderPlanUnknownTargetErrors(t *testing.T) {
	_, _, err := RenderPlan("gcp", NewConfig())
	if err == nil {
		t.Fatal("expected error for unknown target")
	}
}

func TestTargetNamesOrder(t *testing.T) {
	got := TargetNames()
	want := []string{"local", "azure", "aws"}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("TargetNames = %v, want %v", got, want)
	}
}

File: ts/src/deploykit.test.ts
import { describe, expect, it } from "vitest";
import {
  type Config,
  TARGET_NAMES,
  defaultConfig,
  getTarget,
  isSecure,
  renderPlan,
} from "./deploykit.js";

function cfg(): Config {
  return defaultConfig();
}

describe("deploykit core", () => {
  it("default config matches python", () => {
    const c = defaultConfig();
    expect(c.appName).toBe("ledgerrag");
    expect(c.image).toBe("ghcr.io/example/ledgerrag:latest");
    expect(c.replicas).toBe(2);
    expect(c.tls).toBe(true);
    expect(c.secretBackend).toBe("auto");
    expect(c.denyAllEgress).toBe(true);
  });

  it.each(["local", "azure", "aws"])("target %s produces a plan", (name) => {
    const plan = getTarget(name).plan(cfg());
    expect(plan.steps.length).toBeGreaterThan(0);
    expect(Object.keys(plan.manifests).length).toBeGreaterThan(0);
    expect(isSecure(plan)).toBe(true);
    expect(plan.target).toBe(name);
  });

  it("secure defaults pass", () => {
    const plan = getTarget("azure").plan(cfg());
    expect(isSecure(plan)).toBe(true);
    expect(plan.warnings).toHaveLength(0);
  });

  it("insecure config is flagged", () => {
    const c = cfg();
    c.tls = false;
    c.denyAllEgress = false;
    const plan = getTarget("aws").plan(c);
    expect(isSecure(plan)).toBe(false);
    expect(plan.warnings.some((w) => w.includes("TLS"))).toBe(true);
    expect(plan.warnings.some((w) => w.includes("egress"))).toBe(true);
  });

  it("plaintext secret backend is flagged", () => {
    const c = cfg();
    c.secretBackend = "plaintext";
    const plan = getTarget("local").plan(c);
    expect(isSecure(plan)).toBe(false);
    expect(plan.warnings).toHaveLength(1);
    expect(plan.warnings[0]).toContain("plaintext");
  });

  it("warning ordering is stable", () => {
    const c = cfg();
    c.tls = false;
    c.denyAllEgress = false;
    c.secretBackend = "plaintext";
    const plan = getTarget("local").plan(c);
    expect(plan.warnings).toEqual([
      "TLS disabled -- traffic would be unencrypted",
      "egress not restricted -- data exfil risk",
      "secrets stored in plaintext",
    ]);
  });

  it("target selects the right secret backend", () => {
    expect(getTarget("azure").plan(cfg()).manifests["main.bicep"]).toContain("keyvault");
    expect(getTarget("aws").plan(cfg()).steps.join("\n")).toContain("secrets-manager");
    expect(getTarget("local").plan(cfg()).manifests["values.yaml"]).toContain("k8s-secret");
  });

  it("non-auto secret backend override propagates", () => {
    const c = cfg();
    c.secretBackend = "vault";
    const plan = getTarget("local").plan(c);
    expect(plan.manifests["values.yaml"]).toContain("secretBackend: vault");
  });

  it("unknown target throws with exact message", () => {
    expect(() => getTarget("gcp")).toThrowError(
      "unknown target 'gcp'. available: ['local', 'azure', 'aws']",
    );
  });

  it("renders helm values exactly", () => {
    const plan = getTarget("local").plan(cfg());
    expect(plan.manifests["values.yaml"]).toBe(
      "image: ghcr.io/example/ledgerrag:latest\nreplicas: 2\ntls: true\nsecretBackend: k8s-secret\nnetworkPolicy:\n  denyAllEgress: true\n",
    );
  });

  it("renders bicep exactly", () => {
    const plan = getTarget("azure").plan(cfg());
    expect(plan.manifests["main.bicep"]).toBe(
      "// Bicep for ledgerrag\nparam image string = 'ghcr.io/example/ledgerrag:latest'\nparam secretBackend string = 'keyvault'\n",
    );
  });

  it("renders terraform exactly", () => {
    const plan = getTarget("aws").plan(cfg());
    expect(plan.manifests["main.tf"]).toBe(
      '# Terraform for ledgerrag\nvariable "image" { default = "ghcr.io/example/ledgerrag:latest" }\n# secrets: secrets-manager\n',
    );
  });

  it("tls off changes ingress and front steps", () => {
    const c = cfg();
    c.tls = false;
    const azureSteps = getTarget("azure").plan(c).steps.join("\n");
    expect(azureSteps).not.toContain("with TLS");
    expect(azureSteps).toContain("enable ingress");
    const awsSteps = getTarget("aws").plan(c).steps.join("\n");
    expect(awsSteps).not.toContain("ACM TLS cert");
    expect(awsSteps).toContain("front with ALB");
  });

  it("deny egress off removes the network policy step", () => {
    const c = cfg();
    c.denyAllEgress = false;
    const steps = getTarget("local").plan(c).steps;
    expect(steps).toHaveLength(4);
    expect(steps.some((s) => s.includes("NetworkPolicy"))).toBe(false);
  });

  it("renders local steps exactly with defaults", () => {
    const steps = getTarget("local").plan(cfg()).steps;
    expect(steps).toEqual([
      "ensure kind/minikube cluster is running",
      "create namespace ledgerrag",
      "create k8s-secret for app credentials",
      "helm install ledgerrag ./charts/ledgerrag --set image=ghcr.io/example/ledgerrag:latest",
      "apply default-deny NetworkPolicy",
    ]);
  });

  it("render plan secure returns zero", () => {
    const result = renderPlan("local", cfg());
    expect(result.code).toBe(0);
    expect(result.text).toContain("secure-by-default checks: PASS");
    expect(result.text.startsWith("== DeployKit plan: local / ledgerrag ==\n")).toBe(true);
  });

  it("render plan insecure returns one", () => {
    const c = cfg();
    c.tls = false;
    c.denyAllEgress = false;
    const result = renderPlan("aws", c);
    expect(result.code).toBe(1);
    expect(result.text).toContain("SECURITY WARNINGS:");
  });

  it("render plan unknown target throws", () => {
    expect(() => renderPlan("gcp", cfg())).toThrowError();
  });

  it("target names are in order", () => {
    expect([...TARGET_NAMES]).toEqual(["local", "azure", "aws"]);
  });

  it("target name property is exposed", () => {
    expect(getTarget("local").name).toBe("local");
    expect(getTarget("azure").name).toBe("azure");
    expect(getTarget("aws").name).toBe("aws");
  });
});
